lms/djangoapps/commerce/utils.py

```diff
@@ -109,3 +117,261 @@ def upgrade_url(self, user, course_key):
+
+
+def _process_refund(refund_ids, api_client, course_product):
+    """
+    Helper method to process a refund for a given course_product. This method assumes that the User has already
+    been unenrolled.
+
+    Returns:
+        bool: True if the refund process was successful, False if there are any Errors that are not handled
+    """
+    config = CommerceConfiguration.current()
+
+    if config.enable_automatic_refund_approval:
+        refunds_requiring_approval = []
+
+        for refund_id in refund_ids:
+            try:
+                # NOTE: The following assumes that the user has already been unenrolled.
+                # We are then able to approve payment. Additionally, this ensures we don't tie up an
+                # additional web worker when the E-Commerce Service tries to unenroll the learner.
+                api_client.refunds(refund_id).process.put({'action': 'approve_payment_only'})
+                log.info('Refund [%d] successfully approved.', refund_id)
+            except:  # pylint: disable=bare-except
+                # Push the refund to Support to process
+                log.exception('Failed to automatically approve refund [%d]!', refund_id)
+                refunds_requiring_approval.append(refund_id)
+    else:
+        refunds_requiring_approval = refund_ids
+
+    if refunds_requiring_approval:
+        # XCOM-371: this is a temporary measure to suppress refund-related email
+        # notifications to students and support for free enrollments.  This
+        # condition should be removed when the CourseEnrollment.refundable() logic
+        # is updated to be more correct, or when we implement better handling (and
+        # notifications) in Otto for handling reversal of $0 transactions.
+        if course_product.mode != 'verified':
+            # 'verified' is the only enrollment mode that should presently
+            # result in opening a refund request.
+            msg = 'Skipping refund support notification for non-verified mode for user [%s], course [%s], mode: [%s]'
+            course_identifier = course_product.course_id
+            if course_product.uuid:
+                course_identifier = str(course_product.uuid)
+                msg = ('Skipping refund support notification for non-verified mode for user [%s], '
+                       'course entitlement [%s], mode: [%s]')
+            log.info(
+                msg,
+                course_product.user.id,
+                course_identifier,
+                course_product.mode,
+            )
+            return False
+        else:
+            try:
+                return _send_refund_notification(course_product, refunds_requiring_approval)
+            except:  # pylint: disable=bare-except
+                # Unable to send notification to Support, do not break as this method is used by Signals
+                log.warning('Could not send support notification for refund.', exc_info=True)
+                return False
```

Comment:
Just curious, but what happens if the refunds are successfully created in ecommerce but not approved, and we fail to create the Zendesk ticket? Here, we will return `False` and the DB changes revoking the entitlement will be rolled back, but what will happen to those orphaned refund_ids in ecommerce?

Reply:
It will be a refund awaiting approval.  This will sit in limbo until the User contacts support most likely and then they will process.  I can check with Deen, but I would hope we check the Refund dashboard regularly for orphaned refunds like this.

Reply:
What will happen if they retry their request? Will we create another set of new refund ids in ecommerce, or will ecommerce throw an error because we'd be attempting to create duplicate refunds?